plutus-tx/src/PlutusTx/Coverage.hs

```diff
@@ -0,0 +1,170 @@
+    deriving Flat via (AsSerialize Metadata)
+
+instance Pretty Metadata where
+  pretty = fromString . show
+
+newtype CoverageMetadata = CoverageMetadata { _metadataSet :: Set Metadata }
+    deriving (Ord, Eq, Show, Generic, Serialise)
+    deriving newtype (Semigroup, Monoid)
+    deriving Flat via (AsSerialize CoverageMetadata)
+
+makeLenses ''CoverageMetadata
+
+instance Pretty CoverageMetadata where
+  pretty = pretty . Set.toList . _metadataSet
+
+-- | This type keeps track of all coverage annotations and where they have been inserted / what
+-- annotations are expected to be found when executing a piece of code.
+data CoverageIndex = CoverageIndex { _coverageAnnotations :: Set CoverageAnnotation
+                                   , _coverageMetadata    :: Map CoverageAnnotation CoverageMetadata }
+                      deriving (Ord, Eq, Show, Generic, Serialise)
+                      deriving Flat via (AsSerialize CoverageIndex)
+
+makeLenses ''CoverageIndex
+
+instance Semigroup CoverageIndex where
+  ci <> ci' = CoverageIndex (_coverageAnnotations ci <> _coverageAnnotations ci')
+                            (Map.unionWith (<>) (_coverageMetadata ci) (_coverageMetadata ci'))
+
+instance Monoid CoverageIndex where
+  mempty = CoverageIndex mempty Map.empty
+
+-- | Include a location coverage annotation in the index
+addLocationToCoverageIndex :: MonadWriter CoverageIndex m => GHC.RealSrcSpan -> m CoverageAnnotation
+addLocationToCoverageIndex src = do
+  let ann = CoverLocation (toCovLoc src)
+  tell $ CoverageIndex (Set.singleton ann) Map.empty
+  return ann
+
+-- | Include a boolean coverage annotation in the index
+addBoolCaseToCoverageIndex :: MonadWriter CoverageIndex m => GHC.RealSrcSpan -> Bool -> Maybe String -> m CoverageAnnotation
+addBoolCaseToCoverageIndex src b mHeadSym = do
+  let ann = boolCaseCoverageAnn src b
+      meta = case mHeadSym of
+        Just name -> Map.singleton ann $ CoverageMetadata $ Set.singleton $ ApplicationHeadSymbol name
+        Nothing   -> Map.empty
+  tell $ CoverageIndex (Set.singleton ann) meta
+  return ann
+
+boolCaseCoverageAnn :: GHC.RealSrcSpan -> Bool -> CoverageAnnotation
+boolCaseCoverageAnn src b = CoverBool (toCovLoc src) b
+
+-- | Obviously this function computes a GHC.RealSrcSpan from a CovLoc
+toCovLoc :: GHC.RealSrcSpan -> CovLoc
+toCovLoc sp = CovLoc (GHC.unpackFS $ GHC.srcSpanFile sp)
+                     (GHC.srcSpanStartLine sp)
+                     (GHC.srcSpanEndLine sp)
+                     (GHC.srcSpanStartCol sp)
+                     (GHC.srcSpanEndCol sp)
+
+newtype CoverageReport = CoverageReport { _coveredAnnotations :: Set CoverageAnnotation }
+  deriving (Ord, Eq, Show)
+  deriving newtype (Semigroup, Monoid)
+
+makeLenses ''CoverageReport
+
+coverageReportFromLogMsg :: String -> CoverageReport
+coverageReportFromLogMsg = fold . fmap (CoverageReport . Set.singleton) . readMaybe
+
+pprCoverageReport :: CoverageIndex -> CoverageReport -> Doc ann
```

Comment:
My argument is much more simple: having a function `pprX` is much more awkward than having a `Pretty` instance for `X`. So that makes me think that maybe we want a type for `CoverageIndex` and `CoverageReport`. But that seems quite like a "report" to me: indeed the current `CoverageReport` doesn't report all that much without the `CoverageIndex`!

Reply:
@UlfNorell suggested renaming `CoverageReport` to `CoverageData` and bundling them in `CoverageReport` and doing basically what you said here. We can do that, it doesn't cause any issues as far as I can see.

Reply:
Up to you. I guess it wasn't clear to me that the current `CoverageReport` was useful as an independent structure at all, but maybe it is. Up to you, anyway, you're using this stuff.